// @meituan-nocode/vite-plugin-nocode-html-transformer/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.devHtmlTransformer = devHtmlTransformer;
exports.prodHtmlTransformer = prodHtmlTransformer;
/**
 * 开发环境
 * @returns
 */
function devHtmlTransformer(artworkId, imageEnv = 'external') {
    return {
        name: 'vite-plugin-nocode-html-transformer:dev',
        transformIndexHtml(html) {
            // 元素选择器脚本
            const elementSelectorScript = imageEnv === 'internal'
                ? 'https://s3plus.vip.sankuai.com/static-prod01/com.sankuai.mcopilot.nocode.front-files/public/script/element-selector-script.js'
                : 'https://s3plus.meituan.net/mcopilot-pub/nocode-dep/element-selector-script.js';
            return html.replace(/<\/head>/, `
              <script>
                window.onerror = function(message, source, lineno, colno, error) {
                  console.log('window.onerror: ', message, source, lineno, colno, error);
                  window.parent.postMessage({
                    type: 'error',
                    data: {
                      message: message,
                      source: source,
                      lineno: lineno,
                      colno: colno,
                      error: error?.stack || error?.message
                    }
                  }, '*');
                  return true;
                };
                console.error = function(...args) {
                  console.log('console.error: ', args);
                  const message = args.join(' ');
                  window.parent.postMessage({
                    type: 'console-error',
                    data: {
                      message: message,
                    }
                  }, '*');
                  return;
                }
              </script>
              <script type='module' src='./hmr-client.js'></script>
              <script src='https://s3plus.meituan.net/mcopilot-pub/nocode-dep/html2canvas.min.js'></script>
              <script src='${elementSelectorScript}?v=${Date.now()}'></script>
              <script>
                (function () {
                  const script = document.createElement('script');
                  script.src = 'https://s3plus.meituan.net/nocode-external/artwork/nocode-runtime-dev.js?v=' + Date.now();
                  script.onload = function() {
                    window.nocode.init('${artworkId}');
                  };
                  document.head.appendChild(script);
                })();
              </script>
            </head>
              `);
        },
    };
}
/**
 * 生产环境
 * @returns
 */
function prodHtmlTransformer(artworkId) {
    return {
        name: 'vite-plugin-nocode-html-transformer:prod',
        transformIndexHtml(html) {
            // 在 </head> 前插入 Owl 预采集模块
            html = html.replace(/<\/head>/, `
              <script>"use strict";!function(u,d){var t="owl",e="_Owl_",n="Owl",r="start",c="error",p="on"+c,f=u[p],h="addEventListener",l="attachEvent",v="isReady",b="dataSet";u[t]=u[t]||function(){try{u[t].q=u[t].q||[];var e=[].slice.call(arguments);e[0]===r?u[n]&&u[n][r]?u[n][r](e[1]):u[t].q.unshift(e):u[t].q.push(e)}catch(e){}},u[e]=u[e]||{preTasks:[],pageData:[],use:function(e,t){this[v]?u[n][e](t):this.preTasks.push({api:e,data:[t]})},run:function(t){if(!(t=this).runned){t.runned=!0,t[b]=[],u[p]=function(){t[v]||t[b].push({type:"jsError",data:arguments}),f&&f.apply(u,arguments)},u[h]&&u[h]("unhandledrejection",function(e){t[v]||t[b].push({type:"jsError",data:[e]})});var e=function(e){!t[v]&&e&&t[b].push({type:"resError",data:[e]})};u[h]?u[h](c,e,!0):u[l]&&u[l](p,e);var n="MutationObserver",r=u[n]||u["WebKit"+n]||u["Moz"+n],a=u.performance||u.WebKitPerformance,s="disableMutaObserver";if(r&&a&&a.now)try{var i=-1,o=u.navigator.userAgent;-1<o.indexOf("compatible")&&-1<o.indexOf("MSIE")?(new RegExp("MSIE (\\d+\\.\\d+);").test(o),i=parseFloat(RegExp.$1)):-1<o.indexOf("Trident")&&-1<o.indexOf("rv:11.0")&&(i=11),-1!==i&&i<=11?t[s]=!0:(t.observer=new r(function(e){t.pageData.push({mutations:e,startTime:a.now()})})).observe(d,{childList:!0,subtree:!0})}catch(e){}else t[s]=!0}}},u[e].runned||u[e].run()}(window,document);</script>
              </head>
              `);
            // 在 </body> 前插入其他脚本
            html = html.replace(/<\/body>/, `
              <script>
                (function () {
                  const script = document.createElement('script');
                  script.src = 'https://s3plus.meituan.net/nocode-external/artwork/nocode-runtime.js?v=' + Date.now();
                  script.onload = function() {
                    window.nocode.init('${artworkId}');
                  };
                  document.head.appendChild(script);
                })();
              </script>
          </body>
          `);
            return html;
        },
    };
}
